### PR TITLE
cabana: fix chart pixmap size on Hi-DPI screens

### PR DESCRIPTION
**Pull request by pd0wm**
TODO:
- [ ] Fix chart button border being the wrong size
- [ ] Track down segfault when dragging to different screen
- [ ] Fix drag/drop pixmap

image

### CODE REVIEW

**Comment by deanlee** [comment]
no idea why...would setting iconsize after create QGraphicsProxyWidget fix it? seticonsize will force toolbutton to update it's geometry
```
QToolButton *remove_btn = new ToolButton("x", tr("Remove Chart"));
close_btn_proxy = new QGraphicsProxyWidget(chart());
close_btn_proxy->setWidget(remove_btn);
int icon_size = style()->pixelMetric(QStyle::PM_SmallIconSize)
remove_btn->setIconSize(icon_size);
close_btn_proxy->setZValue(chart()->zValue() + 11);
```

**Comment by pd0wm** [comment]
That doesn't help. Seems like `chart_pixmap.setDevicePixelRatio(dpr);` makes the buttons look correct, but I haven't found a way to get the pixmap to draw at the right size after changing the device pixel ratio.

**Comment by deanlee** [comment]
I don't know if this helps: QGraphicsView use the following code to cache background to pixmap,  and paint it to viewport(). 
https://codebrowser.dev/qt5/qtbase/src/widgets/graphicsview/qgraphicsview.cpp.html#3516

File: tools/cabana/chart/chart.cc [hunk 1]
```diff
@@ -599,20 +602,19 @@ void ChartView::startAnimation() {
...
     if (chart_pixmap.isNull()) {
-      const qreal dpr = viewport()->devicePixelRatioF();
       chart_pixmap = QPixmap(viewport()->size() * dpr);
-      chart_pixmap.setDevicePixelRatio(dpr);
       QPainter p(&chart_pixmap);
       p.setRenderHints(QPainter::Antialiasing);
-      drawBackground(&p, viewport()->rect());
+      drawBackground(&p, QRect(viewport()->rect().topLeft(), viewport()->rect().size() * dpr));
       scene()->setSceneRect(viewport()->rect());
       scene()->render(&p);
```
Comment (deanlee) [inline tools/cabana/chart/chart.cc:612]: It is also possible that there is no problem anywhere else, except this line.

Comment (deanlee) [reply tools/cabana/chart/chart.cc:612]: should we set target and source rect  ? https://doc.qt.io/qt-6/qgraphicsscene.html#render

**Comment by deanlee** [comment]
> I was able to reproduce on a non hi-dpi screen by setting `QT_SCALE_FACTOR=2 ./cabana`. So I can confirm this problem also exists on linux.


y,  thanks for the information. I'm testing it on laptop with QT_SCALE_FACTOR now:
Screenshot from 2023-04-14 16-59-33

**Comment by deanlee** [comment]
@pd0wm fixed this issue by set target rect to viewport:

`scene()->render(&p, viewport()->rect());`

I'm checking the issue in drag/drop image now.

**Comment by deanlee** [comment]
fixed the drop image:



```
QPixmap getBlankShadowPixmap(const QPixmap &pm, int extent) {
  QGraphicsDropShadowEffect *e = new QGraphicsDropShadowEffect;
  e->setColor(QColor(40, 40, 40, 245));
  e->setOffset(0, 2);
  e->setBlurRadius(10);

  qreal dpr = pm.devicePixelRatio();
  QPixmap src(pm.size());
  src.setDevicePixelRatio(dpr);
  src.fill(Qt::white);

  QGraphicsScene scene;
  QGraphicsPixmapItem item;
  item.setPixmap(src);
  item.setGraphicsEffect(e);
  scene.addItem(&item);
  QSize target_size = pm.size() + QSize(extent * dpr * 2, extent * dpr * 2);
  QPixmap target(target_size);
  target.setDevicePixelRatio(dpr);
  target.fill(Qt::transparent);
  QPainter p(&target);
  scene.render(&p,QRect({0,0}, target_size / dpr), QRectF(-extent, -extent, item.boundingRect().width() + extent * 2, item.boundingRect().height() + extent * 2));
  return target;
}

static QPixmap getDropPixmap(const QPixmap &src) {
  static QPixmap shadow_px;
  const int extent = 10;
  if (shadow_px.size() != src.size() + QSize(extent * 2, extent * 2)) {
    shadow_px = getBlankShadowPixmap(src, extent);
  }
  QPixmap px = shadow_px;
  QPainter p(&px);
  qreal dpr = src.devicePixelRatio();
  int delta_w = (px.width() - src.width()) / dpr;
  int delta_h = (px.height() - src.height()) / dpr;
  p.drawPixmap(QPoint(delta_w / 2, delta_h / 2), src);
  p.setCompositionMode(QPainter::CompositionMode_DestinationIn);
  p.fillRect(delta_w / 2, delta_h / 2, src.width(), src.height(), QColor(0, 0, 0, 200));
  return px;
}
```

**Comment by deanlee** [comment]
created a PR for these changes: https://github.com/commaai/openpilot/pull/27912

**Pull request closed without merging**